File: experiments/source/ocaml/soft_threshold.ml
open Loom

let[@loom.entry] soft_threshold (threshold : float) (x : Tensor1.t) =
  Tensor1.map
    (fun xi ->
      if xi > threshold then
        xi -. threshold
      else if xi < (-.threshold) then
        xi +. threshold
      else
        0.)
    x

File: experiments/source/python/soft_threshold.py
import loom


@loom.entry
def soft_threshold(threshold: float, x: loom.Tensor1) -> loom.Tensor1:
    def shrink(xi: float) -> float:
        return xi - threshold if xi > threshold else xi + threshold if xi < (-threshold) else 0.0

    return loom.Tensor1.map(shrink, x)
